apps/dashboard/app/models/project_request.rb

```diff
@@ -0,0 +1,47 @@
+# OodAppLink is a representation of an HTML link to an OodApp.
+class ProjectRequest
```

Comment:
I do not think this is a good idea to have another class that represents some intermediate state (i.e., project that hasn't been created yet).

Seems like the model should just validate itself before creating side affects (making a directory and so on).

Reply:
I wasn't able to use a single project model because of the default values on creation for the `icon` and `directory` and the default relationship between id and directory.

In my previous solution, using only the project model, I had to create 2 different ways of saving a project. One for the create flow where `id`, `icon` and `directory` are optional and another for the update flow where all should be populated. I found this solution more complex and difficult to understand.

I can revert back to this solution if you prefer to have a single model.

Reply:
I had to look this up because I wondered how ActiveRecord does this.  Clearly they carry a bit of state in every object, whether it's been saved and so on. Turns out they keep a member variable `@new_record`. 

https://github.com/rails/rails/blob/72b356cdf56b72d36f287b4a72c655607867475e/activerecord/lib/active_record/persistence.rb#L1220

My point being that this is not a new problem to solve surely there must be patterns/things we can do in a single class. 

> I can revert back to this solution if you prefer to have a single model.

Yes please, you can create another pull request if you prefer to have both to compare.

> In my previous solution, using only the project model, I had to create 2 different ways of saving a project. One for the create flow where id, icon and directory are optional and another for the update flow where all should be populated. I found this solution more complex and difficult to understand.

I see what you mean here - well clearly something needs to change. I don't necessarily mind two different flows for `create` and `update`.  (we already have `save` and `update`)

https://github.com/OSC/ondemand/blob/0bc609517f6f916fefbb4578b01eb34ea1e4bd5d/apps/dashboard/app/models/project.rb#L80-L83